### ISSUE TITLE
Incomplete backup

### ISSUE THREAD

**Issue by falsifian** [lehrblogger/OkCupid-Message-Downloader#29]
I downloaded my messages today, and I noticed that some of the threads are incomplete: the last few messages are missing.

In the few examples I looked at, it was always the last three.

What's the easiest way for me to help debug this?

**Comment by joelpurra** [comment]
@falsifian: just to confirm; you mean that within a single "broken" conversation the last three messages are missing, right? About how many threads are broken -- 10% or 50%?

Just performed a full download of messages myself, but seem to have gotten all messages (diffing against the previous time I ran the backup script.) There's always the possibility that you've stumbled upon something new (like a site update) though -- since you are getting *most* messages it *shouldn't* be a trivial error.

You can run `okcmd --debug` to get the raw HTML (as well as a few subsequent extraction/parsing/conversion steps) from each conversation in the first page of messages (assuming there are broken threads on the first page). It's best to pipe the debug output to a file as it tends to be a lot of (too much) data.

- Do you see any surprising results?
- Are all messages downloaded properly?
- Are there messages which are downloaded but not written do disk?

Furthermore:

- Do these broken conversations contain more or longer messages than those which were downloaded correctly?
- It could still be some kind of per user flood protection, if you're actively using the site at the same time. Did you try to log out/close all browser tabs with OKCupid?
- If you're using `okcmd`: did you try to delete the `.okcmd-virtualenv/` folder in the same folder as `okcmd`, then run `okcmd` again?
- If you're using `python okc_arrow_fetcher.py`: which versions of python and BeautifulSoup do you use?
- Can run okcmd again, but save to a different file? Then check the diff and see if you can see any patterns. You could also try this with the `--debug` flag.

**Comment by falsifian** [comment]
Some new observations below. No obvious answers. I can send you the HTML for an example conversation in my browser if you like, or let me know if there's an easy way to dump the full HTML received by the program (the debug output seems to just have the HTML tags it thinks are messages).

Determinism:
* I ran okcmd --debug twice just now, and the output of the two runs is byte-for-byte identical (checked with cmp command).  The debug (stderr) output had some differences, but I think they are timestamps.

Entire threads are missing now:
* However, the output is much smaller than last week's output: 2118 lines compared to 9140. It ismissing many entire conversations that were in my run last week. I can still see these conversations through the web interface.
* It seems to be older threads that are missing mostly. I examined seven threads sort-of picked at random from my entire history, and every one was in today's scrape but not last week's. But poking around more recent messages, most of them are in today's scrape. I couldn't find any that weren't in last week's scrape.

The debug output:
* I don't see anything obviously surprising in the debug output. I looked for a couple of the missing messages there but didn't find them.

Analysis of threads that are only partly downloaded (the original issue):
* When a thread is only partly missing, it's typically the last three messages that are missing. I think I remember seeing an exception to this last week, but it's usually three.
* I looked at some shorter threads (just one or two messages) and none were truncated. (I'm looking at last week's backup, from before entire threads started to go missing.)

Your other questions:
* I wasn't very careful about making sure no OkCupid tabs were open. I can try again from a different IP address if you think it's likely to help. Note that the two scrapes I did today had identical output.
* I didn't notice any individual messages that were corrupted or cut off.
* I'm using okcmd, not okc_arrow_fetcher.py. Between today's two scrapes I deleted the .okcmd-virtualenv directory.

Two other things I noticed, in case they matter:
* For threads with accounts that have since been deleted, the "Sorry, X no longer has an account" is interpreted as a message in its own right.
* When OkCupid shows a button that says "See N older messages", N is often a bit smaller than the number of older messages that actually get shown.

**Comment by joelpurra** [comment]
@falsifian:  diffing with a batch of messages downloaded 1.5+ years ago, I can confirm that I am also missing messages. So far I am missing the messages which are visible when loading a conversation, but the "N older messages" all seem downloaded.

So, the focus should probably be on finding any change in the HTML for the three (probably) most recent messages in a conversation where there are older/hidden messages.

Thanks for noticing the missing messages!

**Comment by joelpurra** [comment]
@falsifian: please see #30 for a possible fix!

The problem seems to have been that OKC serves tricky/malformed HTML that was hard to parse for BeautifulSoup 3. Because of this the three (or so) most recent messages "fell out" of the tag containing the conversation from BS3's point of view. Upgrading to BS4 seemed to fix it.

I also added a *whole* lot more HTML to the debug output, just in case. Please try to download messages again, and check the diff. I'm about to perform a full download myself, and am curios to see how many messages may have been missing due to this bug!

**Comment by falsifian** [comment]
Awesome. I tried #30 and got lots of new messages compared to my previous backups. Glancing at a few threads, I couldn't find anything missing.

**Comment by falsifian** [comment]
So I'd consider this closed if #30 gets merged.

**Comment by joelpurra** [comment]
@falsifian: glad it works! Merged just now. Thanks again =)

**Comment by lehrblogger** [comment]
For what it's worth, my message output with this change seems to be as correct as the last version I have from a couple of years ago. It's hard confirm with a `diff` between OkC got rid of subject lines and put them in the message body, but the files are almost exactly the same size/length.

**Comment by joelpurra** [comment]
> the files are almost exactly the same size/length.

Do you have any files from the past months to compare to?

**Comment by lehrblogger** [comment]
I don't think so, it's been a while since I've run the script on my primary account :-/
